tests/ssg_test_suite/combined.py

```diff
@@ -56,10 +56,10 @@ def _modify_parameters(self, script, params):
         # be tested using the self.profile and we return empty profiles
         # metadata.
         if not params["profiles"]:
-            params["profiles"].append(rule.OSCAP_PROFILE_ALL_ID)
+            params["profiles"].extend(self.profile)
             logging.debug(
                 "Added the {0} profile to the list of available profiles for {1}"
-                .format(rule.OSCAP_PROFILE_ALL_ID, script))
+                .format(self.profile, script))
```

Comment:
It makes sense to default to the profile initially provided, instead of `'(all)'`, but it doesn't to avoid inaccurate results with profiles that select different values than what the test scenario is prepared for.
And unfortunately this is badly documented, I think just in the comment above.

But I guess it could be argued that defaulting to `'(all)' doesn't provide much value when you actually want to test a specific one. Maybe it should skip the test scenario.

I think a workaround for your case would be to set `# profile = xccdf_org....content_profile_cis_level1_server` in the test scenario, as it would be preparing it for the profile value anyway?

Reply:
Maybe it is best to start with documentation on how the existing test suite works currently and in what scenarios various rules and profiles are tested and what the intent of the three CLI invocations are. Maybe that's where the disconnect is. :-)

From my understanding:

 - There are two types of rules in the current CaC system (well, technically a third): 
    1. Rules which depend on a value of a variable e.g., `sudo_add_umask`
    2. Rules which are independent and stand-alone entities. E.g., `sudo_remove_nopasswd`
    3. Rules which depend on one or more other rules being enforced and passing to be passing as well. E.g., something like `aide_build_database` or other related rules which technically depends on `package_aide_installed`.  Sometimes they're written in a way that they pass if the package isn't installed. --- Note that we do have `requires` and `conflicts` in the rule itself (though appear to be used only very rarely such as some `sshd` and `auditd` rules and in the `endpoint_security` rules.
 - Rule two is the easiest: its pass/fail should be strictly independent of profile selection and the values of any and all variables.
     - In reality though, its pass/fail status may be dependent on other external rules and whether or not they've been evaluated.
     - Here, it doesn't/shouldn't matter what invocation path we choose. If we choose a rule invocation, a profile invocation, or a combined invocation, it should behave the same (+/- differences in currently hidden dependencies).
     - So, I'm inclined to ignore this type of rule as it isn't interesting.
 - Rule 1 needs some work under the current system (as I understand it):
     - I do see that there's a `# profiles` selection in the test case itself, in conjunction with a `# platform`.
         - While useful in limited cases, I think having the _value_ of the variables explicitly available is more important. This means:
             - Understanding the default value of variables (for use with `(all)`),
             - Loading and parsing the profile itself (when one was explicitly provided).
             - Jinja to load the current value of the variable into the test case.
             - This ensures the _test_ for the rule gets evaluated in the same environment the _check_ and _remediations_ get built in.
     - Here, like I've been saying above, the _intent_ of the execution (and well, documenting what each scenario is supposed to do) matters most. 
         - For `./test_suite.py rule` based execution there is the `--profile` which, given the above, would suffice.
         - For `./test_suite.py profile` based execution, profile choice is a required argument.
         - For `./test_suite.py combined` based execution, profile choice is again a required argument.
         - IMO, this data is _available_ and hence we should be using it. When I call `test_suite combined <profile>` (with `<profile>` not equal to `ALL` or `(all)`), I shouldn't be evaluating rules and conditions under the `(all)` profile; it should always be under the profile I selected on the CLI. The same principal holds, IMO, elsewhere.
         - For bulk execution, an explicit `ALL` or `(all)` should be used instead.
         - When multiple profiles are selected (with `profile` or `combined` -- note that `rule` doesn't support such), I think we could execute each rule/scenario separately under each profile. My 2c though.
 - Once rule 1 work is done, the test system could be extended to understand the `requires` and `conflicts` options and execute all required dependencies prior to executing the given rule.

Concretely, based on my current understanding, I'd propose:
 - Adding better documentation on _how_ the test suite is organized and the intent between different execution methods.
 - Add Jinja2 support to test cases to avoid the need for `profile` and `platform` as much -- to allow for _scenario_ dependent tests, making them value/platform independent.
 - Clean up profile selection for OSCAP invocations and passing that info into the test case iteslf (for use with jinja macros).
 - Updating the test suite to understand the `requires` (and `conflicts`? I'm not sure) fields. 

Thoughts?

It'll be a good bit of work... I'd like someone with familiarity with the test system as it currently exists to do the documentation... To make sure I'm not missing something obvious :D  But I'd be interested in contributing jinja support, better profile execution understanding, and dependency tracking support eventually if it is of interest the project.